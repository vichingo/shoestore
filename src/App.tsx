import { useState } from "react";

//custom hooks
import {useLocalStorage} from "./hooks/useLocalStorage";

//data
import * as data from "./data/shoes.json";

//context
import { WishlistContext } from "./stores/WishlistStore";

//compontents
import Header from "./components/Header/Header";
import Item from "./components/Item/Item";
import Wishlist from "./components/WishList/Wishlist";
import Drawer from "@material-ui/core/Drawer";
import Grid from "@material-ui/core/Grid";

//types
import { WishlistItemType } from "./App.types";

//styles
import { Wrapper } from "./App.styles";

const App: React.FC = () => {

  const [wishlistOpen, setWishlistOpen] = useState(false);
  const [wishlistItems, setWishlistItems] = useLocalStorage("wishListItems", [] as WishlistItemType[]);

  const products: WishlistItemType[] = data.products;

  const getTotalItems = (items: WishlistItemType[]) => items.reduce((acc: number, item) => acc + item.amount!, 0);

  const handleAddToWishlist = (clickedItem: WishlistItemType) => {
    
    setWishlistItems(prev => {

      // Item already in the cart
      const isItemOnList = prev.find(item => item.id === clickedItem.id)
      
      // Add to item in cart  
      if (isItemOnList) {
        return prev.map(item => 
          item.id === clickedItem.id ? {...item, amount: item.amount! + 1} : item
        );
      }
      
      // Add new item to the cart
      return [...prev, { ...clickedItem, amount: 1 }];
    });
  };

  const handleRemoveFromWishlist = (id: number) =>{
    setWishlistItems(prev =>
      prev.reduce((acc, item) => {
        if (item.id === id) {
          if (item.amount === 1) return acc;
          return [...acc, { ...item, amount: item.amount! - 1 }];
        } else {
          return [...acc, item];
        }
      }, [] as WishlistItemType[])
    );
  };

  return (
    <Wrapper>
      <WishlistContext.Provider value={{wishlistOpen, setWishlistOpen, wishlistItems}}>
        <Header getTotalItems={getTotalItems}/>
      </WishlistContext.Provider>
      <Drawer anchor="right" open={wishlistOpen} onClose={() => setWishlistOpen(false)}>
        <Wishlist 
          wishlistItems={wishlistItems} 
          addToWishlist={handleAddToWishlist}
          removeFromWishlist={handleRemoveFromWishlist}
          />
      </Drawer>
      <Grid container spacing={2}>
        {products?.map((item) => (
          <Grid item key={item.id} xs={12} sm={4}>
            <Item item={item} handleAddToWishlist={handleAddToWishlist} />
          </Grid>
        ))}
      </Grid>
    </Wrapper>
  );
};

export default App;

