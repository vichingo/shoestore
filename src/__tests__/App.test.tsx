import App from '../App';
import { shallow } from 'enzyme';

describe("Testing rendering:", () => {
  it("renders App component without crashing", () => {
    shallow(<App />);
  });
  
  it('App snapshot', () => {
    const component = shallow(<App />);
    expect(component).toMatchSnapshot();
  });
});


