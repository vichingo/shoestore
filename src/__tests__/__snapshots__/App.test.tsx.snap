// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing rendering: App snapshot 1`] = `
<styled.div>
  <ContextProvider
    value={
      Object {
        "setWishlistOpen": [Function],
        "wishlistItems": Array [],
        "wishlistOpen": false,
      }
    }
  >
    <Header
      getTotalItems={[Function]}
    />
  </ContextProvider>
  <WithStyles(ForwardRef(Drawer))
    anchor="right"
    onClose={[Function]}
    open={false}
  >
    <Wishlist
      addToWishlist={[Function]}
      removeFromWishlist={[Function]}
      wishlistItems={Array []}
    />
  </WithStyles(ForwardRef(Drawer))>
  <WithStyles(ForwardRef(Grid))
    container={true}
    spacing={2}
  >
    <WithStyles(ForwardRef(Grid))
      item={true}
      key="1"
      sm={4}
      xs={12}
    >
      <Item
        handleAddToWishlist={[Function]}
        item={
          Object {
            "brand_name": "Jordan",
            "id": 1,
            "image": "images/airjordan-low.jpg",
            "name": "AIR JORDAN LOW",
            "price": Object {
              "original": 109.95,
              "promotional": 109.95,
            },
            "sizes": Array [
              "40",
              "40.5",
              "41",
              "42",
              "42.5",
              "43",
              "44",
              "44.5",
              "45",
              "45.5",
              "46",
              "47",
              "47.5",
              "48.5",
              "49.5",
              "50.5",
              "51.5",
              "52.5",
            ],
            "sku": "JOC12O006-A23",
          }
        }
      />
    </WithStyles(ForwardRef(Grid))>
    <WithStyles(ForwardRef(Grid))
      item={true}
      key="2"
      sm={4}
      xs={12}
    >
      <Item
        handleAddToWishlist={[Function]}
        item={
          Object {
            "brand_name": "Adidas",
            "id": 2,
            "image": "images/adidas-superstar.jpg",
            "name": "SUPERSTAR",
            "price": Object {
              "original": 99.95,
              "promotional": 99.95,
            },
            "sizes": Array [
              "36",
              "38",
              "42",
              "44",
              "46",
              "48",
              "36",
              "37",
              "38",
              "39",
              "40",
              "41",
              "43",
              "44",
              "45",
              "48",
            ],
            "sku": "AD115O0IX-Q12",
          }
        }
      />
    </WithStyles(ForwardRef(Grid))>
    <WithStyles(ForwardRef(Grid))
      item={true}
      key="3"
      sm={4}
      xs={12}
    >
      <Item
        handleAddToWishlist={[Function]}
        item={
          Object {
            "brand_name": "Jordan",
            "id": 3,
            "image": "images/airjordan-1-mid.jpg",
            "name": "AIR JORDAN 1 MID",
            "price": Object {
              "original": 114.95,
              "promotional": 114.95,
            },
            "sizes": Array [
              "44",
              "44.5",
              "45",
              "45.5",
              "47",
              "47.5",
              "48.5",
              "49.5",
              "50.5",
              "51.5",
              "52.5",
            ],
            "sku": "JOC12N001-A18",
          }
        }
      />
    </WithStyles(ForwardRef(Grid))>
    <WithStyles(ForwardRef(Grid))
      item={true}
      key="4"
      sm={4}
      xs={12}
    >
      <Item
        handleAddToWishlist={[Function]}
        item={
          Object {
            "brand_name": "Nike",
            "id": 4,
            "image": "images/airmax-90.jpg",
            "name": "AIR MAX 90",
            "price": Object {
              "original": 139.95,
              "promotional": 139.95,
            },
            "sizes": Array [
              "38.5",
              "39",
              "40",
              "40.5",
              "41",
              "42.5",
              "46",
              "47",
              "47.5",
              "49.5",
            ],
            "sku": "NI112O0BT-Q11",
          }
        }
      />
    </WithStyles(ForwardRef(Grid))>
    <WithStyles(ForwardRef(Grid))
      item={true}
      key="5"
      sm={4}
      xs={12}
    >
      <Item
        handleAddToWishlist={[Function]}
        item={
          Object {
            "brand_name": "Nike",
            "id": 5,
            "image": "images/airmax-90-zw.jpg",
            "name": "AIR MAX 90 - zwart",
            "price": Object {
              "original": 139.95,
              "promotional": 139.95,
            },
            "sizes": Array [
              "38.5",
              "39",
              "40",
              "40.5",
              "41",
              "42",
              "42.5",
              "43",
              "44",
              "44.5",
              "45",
              "45.5",
              "46",
              "47",
              "47.5",
              "48.5",
              "49.5",
            ],
            "sku": "NI112O0BT-Q12",
          }
        }
      />
    </WithStyles(ForwardRef(Grid))>
    <WithStyles(ForwardRef(Grid))
      item={true}
      key="6"
      sm={4}
      xs={12}
    >
      <Item
        handleAddToWishlist={[Function]}
        item={
          Object {
            "brand_name": "Timberland",
            "id": 6,
            "image": "images/adv20-cupsole-modern.jpg",
            "name": "ADV 2.0 CUPSOLE MODERN",
            "price": Object {
              "original": 99.95,
              "promotional": 89.95,
            },
            "sizes": Array [
              "40",
              "41",
              "41.5",
              "42",
              "43",
              "43.5",
              "44",
              "44.5",
              "45",
              "45.5",
              "46",
              "47.5",
              "49",
              "50",
            ],
            "sku": "TI112O00P-K11",
          }
        }
      />
    </WithStyles(ForwardRef(Grid))>
  </WithStyles(ForwardRef(Grid))>
</styled.div>
`;
